# Hyperparameter Tuning using HyperDrive

In [1]:
from azureml.core import Workspace, Experiment
from azureml.core.compute import ComputeTarget, AmlCompute
from azureml.core.compute_target import ComputeTargetException
from azureml.widgets import RunDetails
from azureml.train.sklearn import SKLearn
from azureml.train.hyperdrive.run import PrimaryMetricGoal
from azureml.train.hyperdrive.policy import BanditPolicy
from azureml.train.hyperdrive.sampling import RandomParameterSampling
from azureml.train.hyperdrive.runconfig import HyperDriveConfig
from azureml.train.hyperdrive.parameter_expressions import uniform, choice
import os

Firstly, we require our experiment to be defined within our workspace, the information about the same maybe observed as follows.

In [3]:
ws = Workspace.from_config()
experiment_name = 'HyperdriveRun'

experiment=Experiment(ws, experiment_name)

print('Workspace name: ' + ws.name, 
      'Azure region: ' + ws.location, 
      'Subscription id: ' + ws.subscription_id, 
      'Resource group: ' + ws.resource_group, sep = '\n')

run = experiment.start_logging()

Workspace name: quick-starts-ws-138676
Azure region: southcentralus
Subscription id: aa7cf8e8-d23f-4bce-a7b9-1f0b4e0ac8ee
Resource group: aml-quickstarts-138676


## Dataset

The dataset consists of 70 000 records of patients data in 12 features, such as age, gender, systolic blood pressure, diastolic blood pressure, and etc. The target class "cardio" equals to 1, when patient has cardiovascular desease, and it's 0, if patient is healthy.

The task is to predict the presence or absence of cardiovascular disease (CVD) using the patient examination results.

**Data description**

There are 3 types of input features:

Objective:   *factual information*

Examination: *results of medical examination*

Subjective:  *information given by the patient*

**In train.py**

We download the data from the github repository (https://raw.githubusercontent.com/MonishkaDas/nd00333-capstone/master/starter_file/cardio_train.csv) using TabularDatasetFactory from Azure. We then convert it to pandas dataframe and preprocess it and clean it. The data consisted a few outliers and invalid records(like diastolic pressure,ap_lo was higher than systolic pressure,ap_hi in a few cases). We standardize a few columns ("age", "height", "weight", "ap_hi", "ap_lo") and get dummy values for categorical columns ("cholestrol","gluc").

After cleaning we use 'Logistic Regression' from SKLearn, since it is a classification problem. We tune two hyperparameters, '--C','--max_iter'.


## Hyperdrive Configuration

We firstly need a CPU cluster in order to run our operations on.

In [4]:
# Choose a name for your CPU cluster
cpu_cluster_name = "compute-cluster"

# Verify that cluster does not exist already
try:
    cpu_cluster = ComputeTarget(workspace=ws, name=cpu_cluster_name)
    print('Found existing cluster, use it.')
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_D2_V2',
                                                           max_nodes=4)
    cpu_cluster = ComputeTarget.create(ws, cpu_cluster_name, compute_config)

cpu_cluster.wait_for_completion(show_output=True)


Found existing cluster, use it.
Succeeded
AmlCompute wait for completion finished

Minimum number of nodes requested have been provisioned


## HyperDrive Configuration

**Bandit Policy** <br/>
Defines an early termination policy based on slack criteria, and a frequency and delay interval for evaluation. Bandit policy is based on slack factor/slack amount and evaluation interval. Bandit terminates runs where the primary metric is not within the specified slack factor/slack amount compared to the best performing run. I chose the following parameters for the Bandit Policy slack_factor = 0.001, evaluation_interval=1

**Random Parameter Sampling** <br/>
Random sampling supports discrete and continuous hyperparameters. It supports early termination of low-performance runs. In random sampling, hyperparameter values are randomly selected from the defined search space. I used choice to pass the parameters **--C** ( 0.4, 0.5), **max iter** ( 1000, 1100, 1200, 1300, 1400, 1500 ) to the random Sampler.

**primary_metric_name** </br> 
The primary metric in this case is the **Accuracy**

**primary_metric_goal** </br> Our intention is to find maximize the Primary Metric : Accuracy. This is how we recieve the best model based on metrics.

**max_total_runs**, represents total number runs, in this case is taken as 40

**max_concurrent_runs** is taken as 10

In [6]:
early_termination_policy = BanditPolicy(evaluation_interval=1, slack_factor=0.001)

param_sampling = RandomParameterSampling(
    {
        "--C":uniform(0.4,0.5),
        "--max_iter": choice(1000,1100,1200,1300,1400,1500)
    }
)

if "training" not in os.listdir():
    os.mkdir("./training")

estimator = SKLearn(source_directory='./', 
                entry_script='train.py', compute_target=cpu_cluster)

hyperdrive_run_config = HyperDriveConfig(estimator=estimator,
                                hyperparameter_sampling=param_sampling,
                                policy=early_termination_policy,
                                primary_metric_name="Accuracy",
                                primary_metric_goal=PrimaryMetricGoal.MAXIMIZE,
                                max_total_runs=40,
                                max_concurrent_runs=10)

'SKLearn' estimator is deprecated. Please use 'ScriptRunConfig' from 'azureml.core.script_run_config' with your own defined environment or the AzureML-Tutorial curated environment.


In [7]:
#TODO: Submit your experiment
hyperdrive_run = experiment.submit(config=hyperdrive_run_config)

## Run Details

The `Rundetails` widget, as the name suggests gives us greater insight about how the Run is proceeding, enabling us to monitor and understand the situation, thereby dealing with it accordingly.

In [8]:
RunDetails(hyperdrive_run).show()
hyperdrive_run.wait_for_completion(show_output=True)

_HyperDriveWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO'…

RunId: HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d
Web View: https://ml.azure.com/experiments/HyperdriveRun/runs/HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d?wsid=/subscriptions/aa7cf8e8-d23f-4bce-a7b9-1f0b4e0ac8ee/resourcegroups/aml-quickstarts-138676/workspaces/quick-starts-ws-138676

Streaming azureml-logs/hyperdrive.txt

"<START>[2021-02-12T10:08:15.968487][API][INFO]Experiment created<END>\n""<START>[2021-02-12T10:08:16.405710][GENERATOR][INFO]Trying to sample '10' jobs from the hyperparameter space<END>\n""<START>[2021-02-12T10:08:16.697248][GENERATOR][INFO]Successfully sampled '10' jobs, they will soon be submitted to the execution target.<END>\n"<START>[2021-02-12T10:08:17.3185112Z][SCHEDULER][INFO]The execution environment is being prepared. Please be patient as it can take a few minutes.<END>

Execution Summary
RunId: HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d
Web View: https://ml.azure.com/experiments/HyperdriveRun/runs/HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d?wsid=/subscriptions/aa7cf8e

{'runId': 'HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d',
 'target': 'compute-cluster',
 'status': 'Completed',
 'startTimeUtc': '2021-02-12T10:08:15.642269Z',
 'endTimeUtc': '2021-02-12T10:24:45.767765Z',
 'properties': {'primary_metric_config': '{"name": "Accuracy", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': 'ba3c7f8b-240b-4447-a39e-e83881ddbd6f',
  'score': '0.716566866267465',
  'best_child_run_id': 'HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d_1',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://mlstrg138676.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=B2diFiAYOW7z1JnoOCp7%2FwsIlSSGdSG4ltqkV6rs%2FOY%3D&st=2021-02-12T10%3A14%3A58Z&se=2021-02-12T18%3A24%3A58Z&sp=r'},
 'submittedBy': 'ODL_User 138676

## Best Model

**Best Run ID **:  HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d_1

**Accuracy:**  0.716566866267465

**Metrics: **  

*Regularization Strength:*: 0.40955258766901365

*Max iterations*: 1300, 

**Parameters:** 

*--C* :  0.40955258766901365

*--max_iter* :  1300



In [9]:
best_run = hyperdrive_run.get_best_run_by_primary_metric()
best_run_metrics = best_run.get_metrics()
parameter_values = best_run.get_details()['runDefinition']['arguments']

print('Best Run ID : ',best_run.id)
print('\nMetrics: ',best_run_metrics)

print('\nParameters:', parameter_values)
print('\nAccuracy: ', best_run_metrics['Accuracy'])

Best Run ID :  HD_da36b743-a9e9-40f3-aaa5-d4c5d7fdad9d_1

Metrics:  {'Regularization Strength:': 0.40955258766901365, 'Max iterations:': 1300, 'Accuracy': 0.716566866267465}

Parameters: ['--C', '0.40955258766901365', '--max_iter', '1300']

Accuracy:  0.716566866267465


In [10]:
model = best_run.register_model(model_name='Cardio-HyperDriveModel', model_path='./')